src/EventStore.Core.XUnit.Tests/Scavenge/SoftDeleteTests.cs

```diff
@@ -0,0 +1,203 @@
+﻿using System;
+using System.Threading.Tasks;
+using EventStore.Core.Tests.TransactionLog.Scavenging.Helpers;
+using EventStore.Core.TransactionLog.Scavenging;
+using EventStore.Core.XUnit.Tests.Scavenge.Sqlite;
+using Xunit;
+using static EventStore.Core.XUnit.Tests.Scavenge.StreamMetadatas;
+
+namespace EventStore.Core.XUnit.Tests.Scavenge {
+	public class SoftDeleteTests : SqliteDbPerTest<SoftDeleteTests> {
+		[Fact]
+		public async Task undelete_when_soft_delete_across_chunk_boundary() {
+			// accumulation has to go up to the scavenge point and not stop at the end of the chunk
+			// before, otherwise we could accidentally scavenge the new stream.
+			// the important point is that the scavenge point cant 'appear' to be between the
+			// new records and the new metadata. this could cause problems any time
+			// the scavenge point operates as if it were in the middle of events supposed to bewritten
+			// transactionally.
+			var t = 0;
+			await new Scenario()
+				.WithDbPath(Fixture.Directory)
+				.WithDb(x => x
+					.Chunk(
+						// stream before deletion
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "ab-1"),
+						// delete
+						Rec.Write(t++, "$$ab-1", "$metadata", metadata: SoftDelete),
+						// new write that undeletes the stream, but the metadata lands
+						// in the next chunk
+						Rec.Write(t++, "ab-1"))
+					.Chunk(
+						Rec.Write(t++, "$$ab-1", "$metadata", metadata: TruncateBefore3),
+						ScavengePointRec(t++)))
+				.WithState(x => x.WithConnectionPool(Fixture.DbConnectionPool))
+				.RunAsync(
+					x => new[] {
+						x.Recs[0].KeepIndexes(4),
+						x.Recs[1].KeepIndexes(0, 1),
+					});
+		}
+
+		[Fact]
+		public async Task simple_soft_delete() {
+			var t = 0;
+			await new Scenario()
+				.WithDbPath(Fixture.Directory)
+				.WithDb(x => x
+					.Chunk(
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "$$ab-1", "$metadata", metadata: SoftDelete))
+					.Chunk(ScavengePointRec(t++)))
+				.WithState(x => x.WithConnectionPool(Fixture.DbConnectionPool))
+				.AssertState(state => {
+					Assert.False(state.TryGetOriginalStreamData("ab-1", out _));
+					Assert.False(state.TryGetMetastreamData("$$ab-1", out _));
+				})
+				.RunAsync(
+					x => new[] {
+						x.Recs[0].KeepIndexes(2, 3), // keep the last event
+						x.Recs[1],
+					});
+		}
+
+		[Fact]
+		public async Task soft_delete_and_recreate() {
+			var t = 0;
+			await new Scenario()
+				.WithDbPath(Fixture.Directory)
+				.WithDb(x => x
+					.Chunk(
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "$$ab-1", "$metadata", metadata: SoftDelete),
+						Rec.Write(t++, "ab-1"),
+						Rec.Write(t++, "$$ab-1", "$metadata", metadata: TruncateBefore3))
```

Comment:
should we also add a test for the case where the stream is partly recreated i.e the event is written but not the new metadata? (e.g due to a crash mid way) 

in this case, should scavenge delete or keep the new event? at first glance, it sounds like it should keep it because the client might retry the write, resulting in an idempotent write while a scavenge is happening in parallel.

Reply:
it seems that the simplest way to fix it is to convert the metadata's truncate before during accumulation before saving it in the state. it it's equal to EventNumber.DeletedStream, then the last event number before the soft deletion event's position can be fetched from the index.

Reply:
as pointed out by @timothycoleman, the writer checkpoint is moved only after writing both events, so it is atomic. however, the issue could occur if the first event is written exactly at the end of a chunk and the second one fails to be written in the next chunk - at the moment it's rare enough to not need to worry about this case.